### PR TITLE
[5.2] TinyMCE 7

### PR DESCRIPTION
**Pull request by brianteeman**
Pull Request for Issue # .

### Summary of Changes
The majority of changes with 7.x do not impact us as they are for the premium version and the premium plugins. Other changes are really just bug fixes and accessibility improvements.

The full changelog for the open source release of tinymce can be found [here](https://www.tiny.cloud/docs/tinymce/latest/changelog/)

Set tinymce to be used under the gpl licence. https://www.tiny.cloud/docs/tinymce/latest/license-key/#use-tinymce-with-the-gplv2-license - without this change a console warning is displayed

The one change that does need testing is sandbox_iframes

### Testing Instructions
1. Test all editor-xtd buttons 

2. Test the source code editor 
image

3. Test any third party plugin eg https://github.com/brianteeman/slashtools
This uses one of the few breaking changes so should fail https://www.tiny.cloud/docs/tinymce/latest/migration-from-6x/#the-autocompleter-ch-configuration-property-has-been-removed-use-the-trigger-property-instead I will make a new release of it but I am not aware of any other plugin using this.

4. Test drag and drop images

5. Test content templates
image

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by bembelimen** [comment]
Hey @brianteeman , not sure if this answers your question, but the idea for the ```sandbox_iframes``` parameter was: when we go to TinyMCE 7 it should still stay at "true", so all entries get sandboxed by default (and the plugin parameter we added to disable it could be removed) but additionally the current domain should be allowed via: [```sandbox-iframes-exclusions```](https://www.tiny.cloud/docs/tinymce/latest/content-filtering/#sandbox-iframes-exclusions).
Also an extra field should be added, where people can add more domains to allow them.

One concern is, how to prevent, that people open the joomla installation with another domain and add iframes for this domain (as it's then allowed automatically).

**Comment by brianteeman** [comment]
Thanks, that answers most of it I think. Will add the exclusion field

**Comment by dgrammatiko** [comment]
@brianteeman @bembelimen according to their docs you shouldn't expose the switch to turn on/off the sandboxing as an empty array is doing exactly that: `To enable [sandboxing](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#sandbox) on all iframes, set this option to an empty array [].`

In sort the `sandbox_iframes_exclusions` should be an array with default value the value of tinymce (is given in the link above): 

```js
[
  'youtube.com',
  'youtu.be',
  'vimeo.com',
  'player.vimeo.com',
  'dailymotion.com',
  'embed.music.apple.com',
  'open.spotify.com',
  'giphy.com',
  'dai.ly',
  'codepen.io'
]
```

**Comment by brianteeman** [comment]
yes I know. I am still working on it (or will be tonight when care duties end) or you can submit a pr to this draft branch ;)

**Comment by Quy** [comment]
Should it be only the domain name without `http`?

**Comment by brianteeman** [comment]
> Should it be only the domain name without `http`?

??

**Comment by Quy** [comment]
These are listed without `http`. Should the additional values be host names or domain names (without `http`)?

```
[
  'youtube.com',
  'youtu.be',
  'vimeo.com',
  'player.vimeo.com',
  'dailymotion.com',
  'embed.music.apple.com',
  'open.spotify.com',
  'giphy.com',
  'dai.ly',
  'codepen.io'
]
```

**Comment by brianteeman** [comment]
> These are listed without `http`. Should the additional values be host names or domain names (without `http`)?
> 
> ```
> [
>   'youtube.com',
>   'youtu.be',
>   'vimeo.com',
>   'player.vimeo.com',
>   'dailymotion.com',
>   'embed.music.apple.com',
>   'open.spotify.com',
>   'giphy.com',
>   'dai.ly',
>   'codepen.io'
> ]
> ```

The current domain and this list of common domains are excluded by default.

**Comment by Quy** [comment]
Yes, I understand. but when adding new ones to the exclusions list, there is a hint `https://` so maybe it can interpreted to prefix it with `https://`.

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 1]
```diff
@@ -410,6 +410,21 @@ public function prepareTinyMCEUploadPath(string $path): string
...
         }
 
+        // Add the current domain to the sandbox_iframes_exclusions list
+        $sandboxIframesExclusions = Uri::root();
+
+        // Build the list of additional domains to add to the sandbox_iframes_exclusions list
+        if (isset($extraOptions->sandbox_iframes_exclusions) && $extraOptions->sandbox_iframes_exclusions) {
+            $exclusionsArray = [];
+            foreach ($extraOptions->sandbox_iframes_exclusions as $value) {
+                if (isset($value->exclusion_url)) {
```
Comment (Quy) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: I assume there would be a value at this point so no need to do the check?

**Comment by brianteeman** [comment]
I understand you now. Thanks for your perseverance I will look into it in the morning

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 2]
```diff
@@ -45,25 +45,30 @@ PLG_TINY_FIELD_HTMLWIDTH_LABEL="HTML Width"
...
 PLG_TINY_FIELD_NUMBER_OF_SETS_LABEL="Number of Sets"
 PLG_TINY_FIELD_PASTE_AS_TEXT_LABEL="Paste As Text"
 PLG_TINY_FIELD_PATH_LABEL="Element Path"
 PLG_TINY_FIELD_PROHIBITED_LABEL="Prohibited Elements"
 PLG_TINY_FIELD_RESIZE_HORIZONTAL_LABEL="Horizontal Resizing"
 PLG_TINY_FIELD_RESIZING_LABEL="Resizing"
+PLG_TINY_FIELD_SANDBOX_IFRAMES_DESC="This is a security feature that restricts what the iframe can do by setting a sandbox attribute for each iframe. It is recommended to enable this feature for security reasons. Only disable if the iframe is not loading and you're sure that the iframed content is safe."
+PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_DESC="Additional URL host names can be added to be excluded from iframe sandboxing. The current domain and this list of common domains are excluded by default. 'youtube.com', 'youtu.be', 'vimeo.com', 'player.vimeo.com', 'dailymotion.com', 'embed.music.apple.com', 'open.spotify.com', 'giphy.com', 'dai.ly', 'codepen.io'."
+PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_LABEL="Sandbox Iframe Exclusions List"
+PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSION_DOMAIN_LABEL="URL Host Name"
```
Comment (Quy) [inline administrator/language/en-GB/plg_editors_tinymce.ini:63]:
```suggestion
PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSION_DOMAIN_LABEL="Domain Name"
```

File: plugins/editors/tinymce/forms/setoptions.xml [hunk 3]
```diff
@@ -238,7 +238,23 @@
...
+			name="sandbox_iframes_exclusions"
+			type="subform"
+			label="PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_LABEL"
+			description="PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_DESC"
+			multiple="true"
+			showon="sandbox_iframes:1"
+			>
+			<form>
+				<field
+					name="exclusion_url"
```
Comment (Quy) [inline plugins/editors/tinymce/forms/setoptions.xml]: Rename variable?

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 4]
```diff
@@ -410,6 +410,21 @@ public function prepareTinyMCEUploadPath(string $path): string
             }
         }
 
+        // Add the current domain to the sandbox_iframes_exclusions list
+        $sandboxIframesExclusions = Uri::root();
```
Comment (Quy) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: Returns `http://localhost/Joomla_5.2.0-alpha4-dev+pr.43909-Development-Full_Package/`

File: administrator/language/en-GB/plg_editors_tinymce.ini (hunk 2 above)
Comment (Quy) [inline administrator/language/en-GB/plg_editors_tinymce.ini]:
```suggestion
PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_DESC="Additional domain names can be added to be excluded from iframe sandboxing. The current domain and this list of common domains are excluded by default. 'youtube.com', 'youtu.be', 'vimeo.com', 'player.vimeo.com', 'dailymotion.com', 'embed.music.apple.com', 'open.spotify.com', 'giphy.com', 'dai.ly', 'codepen.io'."
PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSIONS_LIST_LABEL="Sandbox Iframe Exclusions List"
PLG_TINY_FIELD_SANDBOX_IFRAMES_EXCLUSION_DOMAIN_LABEL="Domain Name"
```

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php (hunk 4 above)
Comment (brianteeman) [reply plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: that is what I would expect. If that is nto what is needed to then I need some help

Comment (Quy) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]:
```suggestion
        $sandboxIframesExclusions = Uri::getInstance()->getHost();
```

**Comment by brianteeman** [comment]
Note to self
Review the skins and themes https://www.tiny.cloud/docs/tinymce/latest/customize-ui/

**Comment by HLeithner** [comment]
I rebase this for now to 5.3 (actually 6.0 should be the target), since it's late in the cycle and actually it might be a b/c break if tinymce follows semver. If maintainers/production has finally discussed this a rebase to 5.2 is possible.
So please don't update the branch to 5.3 code

**Comment by brianteeman** [comment]
@HLeithner I specifically asked which version to make this for and was told https://github.com/joomla/joomla-cms/issues/43110#issuecomment-2274042689 - please make up your minds

**Pull request closed without merging**